renderer_opengl: add PICA->GLSL shader decompiler

wwylele opened the pull request:
This is the core part of #3499. I extracted out so that it can be reviewed first. In this PR, I only added the source file, without adding any users, so that we can focus on the implementation first. Other parts in #3499 still have some unclear issue including AMD problems, which will be investigated after this.

Here is a doc about the design

----

## Concepts

A subroutine is a collection of code. Each subroutine is identified by two code addresses: the entry point (named "begin" in the code) and the return point (named "end" in the code). The code execution starts from the entry point. When the execution reaches the return point, the subroutine exits and the execution goes back to the caller. However, the entry point is not necessarily the lower boundary of the code belonging to the subroutine; the return point is not necessarily the upper boundary, either.

A subroutine is also characterized by one property called "exit method", which can be one of the following three values:
 - Always END,
 - Always return,
 - Either END or return conditionally.

After entering the subroutine, the execution flow would either terminates the entire shader program by executing the END instruction (including an END in a inner level subroutine), or return to the caller by reaching the return point. The "exit method" property defines by which way code paths would exit the subroutine.

A new subroutine execution is initiated by one of the following flow control instructions: CALL(U/C), IF(U/C) and LOOP. The entry point and the return point are both specified by the instruction. Note that IF(U/C) initiates two subroutines for both "if" and "else" branches.

The "main" program is also a special subroutine, which initiates all other subroutines. The "main" subroutine is defined with entry_point = main_offset, and return_point = MAX_PROGRAM_CODE_LENGTH. The return point is only conceptional and is never expected to be reachable. Therefor the "main" program should always have "Always END" as its exit method.

JMP(U/C) instructions don't initiate new subroutine. Instead, the branching code of the JMP(U/C) is treated in the same subroutine

## Decompilation Process

The process consists of two steps.

### Control Flow Analysis

In this pass, the shader code is traversed (depth-first) from the main_offset. During the process, exit methods are determined for each entry point, subroutines are identified, and labels for JMP are collected. The analysis function is executed recursively on each subroutine call / branching point, and the exit methods from each branch are combined to feed back to the caller. Note that although JMP(U/C) instructions are not considered as subroutine initiating instruction, they are still conceptually analysed by entry point and return point (= same return point of the beloning subroutine).

One difficulty in this pass is the implicit loop by backward branching. Only CALL(U/C) and JMP(U/C) are assumed able to do this. Loop by CALL(U/C) would actually form a recursion, which is assumed never happening and ASSERTed in this pass.

Loop by JMP(U/C) is handled by an intermediate forth "exit method" value "Undetermined". This value is first cached into the table, and got overridden by a definite value on subroutine exit. If a value fetched from the cache is "Undetermined", it means the branching point forms a loop. There are a few assumptions about the loop:
 - A loop never goes across different subroutines
 - There is no infinite loop

Baesd on this two assumptions, the exit method of a subroutine would be the same as the one without loop-back branch, so the exit method at the branching point, which is normally decided by "merging" the methods of the two branches, is now handled by ignoring the "Undetermined" branch. This would possibly leave an incorrect exit method result at the branching point. However, we eventually only care about the exit method of subroutine entries, whereas the loop branching point is always a JMP instruction, which never initiates a subroutine.

### Code Generation

Each subroutine is compiled to a function in GLSL. These functions have a signature form `bool sub_<start>_<end>();`. The return value indicates whether an END instruction has been executed inside the subroutine. Function callers would cascade the `true` return to the most upper level in order to terminate the entire program.

In a subroutine, Arithmetic/MultiplyAdd instructions are translated to the corresponding GLSL formula. IF(U/C) instructions are compiled to `if(...){...}else{...}` blocks, and LOOP instructions are compiled to `for(...){...}` blocks. CALL(U/C) instructions are compiled to function calls, wrapped by `if` if conditional, and then cascade the return when needed.

JMP instructions are handled like this: the entire function is wrapped by `uint jmp_to = <start>; while(true){switch(jmp_to){...}}`. On each JMP instruction, `jmp_to` is set to the destination address and `break` the `switch` block to do the next iteration of the loop. Each potential JMP destination is labelled by `case <address>:...` to catch the new destination in `jmp_to`. If the current subroutine doesn't have any JMP, then the `while`+`swtich` wrapper is omitted.

The final code output is not a complete vertex shader or a geometry shader; instead, one needs to add definition and data transferring for input/output registers. For PICA geometry shader, one also needs to define `emit()` and `setemit()` functions.

----

The original code is written by @Phanto-m , and I read the code and understood it on my own (without talking with them, unfortunately, because they are not available these days). I made some changes to the original code, including:
 - The original code is one bulk function with all other functions as lambdas inside. I moved them out and made them normal functions, with documentation of each function.
 - The control flow analysis in the original code effectively had two passes: one for finding END instructions and one for collecting subroutines. It also had a lot of additional code traversal for detecting recursion and "Conditional END". I managed to simplify them into only one traversal, one pass.
 - The original code would inline some small subroutines and JMPs in certain conditions. I removed the inline mechanism to simplify the code logic. No performance regression reported so far, and OpenGL driver should be able to inline functions anyway. And after reworking the control flow analysis, I think it wouldn't be hard to add back inline mechanism if we really need it.

----

There are some unresolved issues in the code, including:
 - [ ] Accurate multiplication having undefined behaviour. (has [a potential solution](https://github.com/citra-emu/citra/pull/3499#discussion_r175734697) in #3499 by @degasus,~~I haven't check it yet ~~ but it is not perfect either) (edit: the undefined behaviour is that the current implementation relies on that `mix` will output NaN when it should, but OpenGL says for built-in function NaN is nor required to output. So far most OpenGL implementation does will, and only one user reported the problem)
 - [ ] Inaccurate result on edge cases for comparison function (MIN/MAX/SLT/SGE) ([found by @Tilka](https://github.com/citra-emu/citra/pull/3499#issuecomment-374589454) in #3499). It worth noticing that a similar problem also presents in the existing shader interpreter. I wonder if this is really that important.
 - [ ] Silently hiding out-of-boundary access for indexed uniform registers (#2443) with undefined behaviour. I am not sure what is the desire way to handle this. Which should we do among these two options?
   - clamping the index to make it at least have defined behaviour, without error/warning. This would be just a one-line change.
   - reporting the error from shader. Note that this error can be only detected at shader runtime. I don't know how to throw runtime exceptions / assert / crash the shader at runtime.
   -  (Update: added clamp)

These problems are all about the code generation and wouldn't affect the main structure of the code. I think we can review the main part of the code and try to resolve these issues at the same time.

----

Comments about the emulation accuracy: The GLSL shader path would probably have the lowest accuracy to emulate PICA shader, as it has the least freedom to control the result on edge cases involving Inf, NaN and precision-mismatch numbers. But I think it has its own value, as it is potentially the fastest path that is still accurate enough for most games, which provides a good user experience. We aren't going to remove the old shader emulation on CPU, which has more space for potentially more accurate emulation, and we will continue maintaining all of them.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/3567)
<!-- Reviewable:end -->

wwylele commented:
[Here](https://pastebin.com/8F24H8RF) is a sample GLSL code generated by this decompiler.

degasus commented:
> Other parts in #3499 still have some unclear issue including AMD problems

There is a huge chance that the issues on AMD GPUs are within this PR.

wwylele commented:
@degasus According to @darkf AMD issue is about buffer sync issue and can be workarounded by adding `glFInish` at some places, so I think the shader itself doesn't have problem.

degasus commented on src/video_core/renderer_opengl/gl_shader_decompiler.h:
```diff
@@ -0,0 +1,27 @@
+
+std::string GetCommonDeclarations();
+
+std::string DecompileProgram(const ProgramCode& program_code, const SwizzleData& swizzle_data,
```
I think the interface need some way to return if the decompilation was successful. Maybe define "" as invalid, or add a bool.

MerryMage replied:
`boost::optional<std::string>`, perhaps

degasus commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,911 @@
+    std::set<u32> labels;   ///< Addresses refereced by JMP instructions.
+
+    bool operator<(const Subroutine& rhs) const {
+        if (begin == rhs.begin) {
```
return std::tie(begin, end) < std::tie(rhs.begin, rhs.end);

MerryMage commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,911 @@
+    }
+
+    u32 begin;              ///< Entry point of the subroutine.
+    u32 end;                ///< Return point of the subroutine.
```
If I understand correctly an exception to this would be the main subroutine, for which `end` = `PROGRAM_END`?

wwylele replied:
yes

MerryMage commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,911 @@
+            UNREACHABLE();
+            return "";
+        }
+    };
```
Stray `;`. (Some more further down.)

jroweboy commented:
@degasus oh i sent you this on IRC but i forgot to check if you were online, so you might not have received it.

darkf and I have been trying to hunt down the amd issue, and it seems that it has something to do with streaming buffers. darkf did what you recommended and put glfinish around places till it worked
and the magic glfinish call is right before glDrawRangeElementsBaseVertex. of course this kills the speed, and its like twice as slow as the cpu shader jit now :P so i went and moved the gpu shaders to your new buffer storage code that you wrote, https://github.com/degasus/citra/tree/vertex_streaming but amd still crashes with it. i'm kinda hoping that you might have some insight into this :)

It may not have something to do with streaming buffers as well but I did want to fill you in on what we've found out so far.

degasus commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,911 @@
+            }
+            case OpCode::Id::IFU:
+            case OpCode::Id::IFC: {
+                auto& if_sub = AddSubroutine(offset + 1, instr.flow_control.dest_offset);
```
I wonder about the name "if_sub" and "else_sub" here. The former is executed if the branch is not taken.

wwylele replied:
The names are based on this coding form
```
if (condition) {
    // "if" block here
} else {
    // "else" block here
}
```
So yeah strictly speaking the branching is to the `else` block. 
If this is not intuitive enough for you, do you have some better names to suggest?

Subv commented:
@degasus @wwylele Regarding accessing invalid uniforms, a hwtest i performed a long time ago seems to indicate that doing so crashes the GPU (https://github.com/citra-emu/citra/issues/2443#issuecomment-306986764) however I don't trust these results anymore and they should be re-checked

degasus commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,911 @@
+    }
+
+    /// Scans a range of code for labels and determines the exit method.
+    ExitMethod Scan(u32 begin, u32 end, std::set<u32>& labels) {
```
Do you think there should be an integer for the stack depth to detect recursion?

wwylele replied:
Oops the I did have an ASSERT in `AddSubroutine` for recursion detection (by asserting new subroutines doesn't have an `Undetermined` exit method), but I removed it by mistake at some point. Will add it back.

wwylele commented:
Addressed comments. 

The error handling is internally handled by throwing and catching exceptions, as I don't want to add error handling boilerplate everywhere in all functions for these cold error paths. Though it is a not like our general code style, I don't see any disadvantages using exceptions here.

lioncash commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,925 @@
+            throw DecompileFail("Program does not always end");
+    }
+
+    std::set<Subroutine> GetSubroutines() {
```
This should likely be named something else. This isn't simply acting as a basic getter (ditto where relevant elsewhere)

lioncash commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -0,0 +1,925 @@
+// Refer to the license.txt file included.
+
+#include <exception>
+#include <set>
```
`<map>`, `<tuple>`, and`<utility>` also needs to be included.

degasus commented:
@jroweboy I answered on IRC about this AMD issue a few minutes later, no clue if this was relyed to discord through. Which call crashes with lots of glFinish? The glFinish before the draw call or the glFinish after the draw call? I still doubt the GPU reset is within the driver. It must be something within the shader, maybe an infinite loop, maybe a compiler bug generating broken code. In both cases, you should look at the bound shaders of the crashing glDraw call.

jroweboy commented:
@degasus ah. never arrived. I wondered why things were so quiet from the IRC side lately; the bridge is probably down. I'll check that with @darkf as soon as we get a chance, thanks!

wwylele commented:
Added clamp on indexed uniform access. The MUL issue still doesn't get a better solution (as what degasus suggested is not perfect either). I think we can merge this first if there is no general design problem, and these small points can be fixed later if needed.

wwylele commented:
Also about the AMD issue: from recent experiment it looks like it has something to do with bugs in AMD's separate shader program, which isn't introduced in this PR.

jroweboy commented on src/video_core/renderer_opengl/gl_shader_decompiler.cpp:
```diff
@@ -317,8 +317,10 @@ class GLSLGenerator {
-                index_str +=
-                    std::string(" + address_registers.") + "xyz"[address_register_index - 1];
+                // The clamp is to prevent out-of-bound access, which has been seen in some games
+                // such as Mario Kart 7.
```
Do you have any evidence that this happens in the GPU shaders? When testing AMD issues, I tested clamping, and found that at least on my nvidia card, I never had any out of bounds accesses. I know that our past research into out of bound uniform accesses in the CPU shaders lead us to believe that it should affect GPU shaders too, but I haven't *actually* seen it happen in games that we knew it should.

wwylele replied:
I don't have and I almost disproved myself just now... so what is your opinion? Revert this clamp or just remove the comment?

jroweboy replied:
I'm in favor of removing the clamp from here, as it'll likely just hide bugs and make it harder to track down later.
